### PR TITLE
[Bifrost] read correctly handles seal and tail instability

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] read correctly handles seal and tail instability

- Segments return the known tail_lsn which gets propagated to downstream components.
- Renaming `read_opt()` to replace `read()` and making the non-blocking `read()` the de-facto way to read individual records. Readstream will be the only way for tailing readers.
- Implements the intricate handling of `read()` during in-flight reconfiguration. This now requires bifrost to determine a safe known tail before allowing the `read()` to take place.
- Testing read single record cross segments

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1725).
* #1726
* __->__ #1725
* #1722
* #1720
* #1718
* #1717
* #1705
* #1704

### CODE REVIEW

File: crates/types/src/logs/metadata.rs [hunk 1]
```diff
@@ -222,27 +229,36 @@ impl Chain {
         // NOTE: Hopefully at some point we will use the nightly Cursor API for
```
Comment (tillrohrmann) [inline crates/types/src/logs/metadata.rs:229]: Rustdoc for `find_segment_for_lsn` seems outdated.

**Review by tillrohrmann — APPROVED** [review]
Great work! LGTM. +1 for merging.

File: crates/bifrost/src/bifrost.rs [hunk 2]
```diff
@@ -283,28 +275,73 @@ impl BifrostInner {
...
+                None => {
+                    // optimization to avoid find_tail if we know it's safe to perform this read.
+                    if let Some(known_unsealed_tail) = loglet.last_known_unsealed_tail() {
+                        if known_unsealed_tail > from {
+                            // safe to proceed without find_tail.
+                            break (known_unsealed_tail, loglet);
+                        }
+                    }
+
+                    let tail = loglet.find_tail().await?;
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/bifrost.rs:308]: Most likely yes, that's why reading with a readstream will be better (more details on this in future PRs)

File: crates/types/src/logs/metadata.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/types/src/logs/metadata.rs:229]: Is it correct now?

Comment (tillrohrmann) [reply crates/types/src/logs/metadata.rs:229]: 👍

**Pull request merged**